moped-editor/src/views/projects/projectView/ProjectNameComplex.js

```diff
@@ -0,0 +1,292 @@
+
+  const DEFAULT_SNACKBAR_STATE = {
+      open: false,
+      message: null,
+      severity: "success",
+    };
+
+  // used to control the snackbar's contents and visibility
+  const [snackbarState, setSnackbarState] = useState(DEFAULT_SNACKBAR_STATE);
+
+  // apollo hook returning the async function to update the project names
+  const [updateProjectNames] = useMutation(UPDATE_PROJECT_NAMES_QUERY);
+
+
+
+  // this function is fired when the user clicks on the check mark to save the names
+  const handleAcceptClick = (e) => {
+    e.preventDefault();
+
+    const projectName = document.getElementById("project_name").value;
+    const secondaryName = document.getElementById("secondary_name").value;
+
+    // Check if the primary name is defined and has meaningful content
+    if (!projectName || projectName.trim() === "") { // if it does not, set the error state
+      setPrimaryTitleError(true);
+    } else { // and if it passes the test, clear any error state and save it via a mutation
+      setPrimaryTitleError(false);
+      updateProjectNames({
+        variables: {
+          projectId: props.projectId,
+          projectName: projectName,
+          projectNameSecondary: secondaryName,
+        },
+      })
+      .then((res) => { // if the mutation is successful, show a success snackbar
+        setSnackbarState({
+          open: true,
+          message: <span>Success! The project names have been updated!</span>,
+          severity: "success",
+        });
+      }) 
+      .catch((error) => { // and if it fails, show an error snackbar
+        setSnackbarState({
+          open: true,
+          message: <span>Error. The project names have not been updated.</span>,
+          severity: "error",
+        });
+        })
+        .finally(() => { // return to the view mode, clear the snackbar after 3 seconds, 
+                         // and alert the parent component of the change
+        props.setIsEditing(false);
+        setTimeout(() => setSnackbarState(DEFAULT_SNACKBAR_STATE), 3000);
+        props.updatedCallback();
+      });
+    }
+  };
+
+  // this is fired when the user clicks on the 'X' to cancel the edit
+  const handleCancelClick = (e) => {
+    e.preventDefault();
+    props.setIsEditing(false);
+  };
+
+  // this is fired onChange of the project name field to check if it's valid
+  const handleProjectNameChange = (e) => {
+    const projectName = document.getElementById("project_name").value;
+    if (!projectName) {
+      setPrimaryTitleError(true);
+    } else { 
+      setPrimaryTitleError(false); }
+  }
+
+  // this is fired when the snackbar's close button is clicked or by timeout
+  const handleSnackbarClose = (event, reason) => {
+    setSnackbarState(DEFAULT_SNACKBAR_STATE);
+  };
+
+  // Because the layout is non-trivially different between view and edit modes,
+  // we're going to use two different JSX blocks to render the component's content.
+  const viewModeJSX = (
```

Comment:
I hear what you're saying, and some semblance of the approach you're talking about is what I started to do initially but started to run into complications.

I'll try to describe some more detail about my rational: the way our old component worked is that it big blocks of conditions in its JSX flipping between edit mode and view mode, with both modes in one component, so that made the conditional JSX pattern feel in-bounds. Also, I think to break out the input fields here into components, I'm going to need to do the view and edit separately as two components, at least for the secondary name, because the larger layout of the name system visually for view / edit modes is different. I guess you could do it with a prop to control if it shows the edit mode or the view mode and to hide the other, and then use it twice in the page .. but that feels contorted and overly complicated for little benefit, I think ..  like you said, there's a million ways to slice it.

I got to talk to Chia about this briefly earlier, but I don't think I explained the nuance to why I chose to do it this way, or even explained it too clearly here, but not for lack of trying.

If you want to sync on this, I'd love to talk about approaches in handling this in advance of making a revision to this PR, but right now, I still kind of think this slightly monolithic setup is OK given the issue's requirements. But, I do, 100%, recognize what you're saying that this is not ideal. The point you made about how the name I picked indicated that this is too complex and that there are better ways to break this up was verrrry well said. :100: I just don't see the subtleties yet and need some help.

Reply:
Cool, I hear what you are saying too. I'd love to pair on this some today to flesh out what I'm saying more if you have time. Might be easier to get what I'm trying to say across in real time!